File: DSA/PranyaGupta_500126032/Day36Question2.py
class Stack:
    def __init__(self):
        self.stack = []
    def push(self, data):
        self.stack.append(data)
    def pop(self):
        if not self.isEmpty():
            return self.stack.pop()
        else:
            return ''    
    def peek(self):
        if not self.isEmpty():
            return self.stack[-1]
        else:
            return None
    def isEmpty(self):
        return len(self.stack)==0
            
    def display(self):
        print(self.stack) 

class Solution():
    def balancedParanthesis(self, s):
        stk = Stack()
        p = {')': '(', ']': '[', '}': '{'}
        for c in s:
            if c in '([{':
                stk.push(c)
            elif c in '})]':
                if stk.isEmpty() or stk.peek()  != p[c]:
                    return "Not Balanced"
                stk.pop()
        if stk.isEmpty():
            return "Balanced"
        else:
            return "Not Balanced{}"             

def main():
    s = input("Enter the parenthesis: ")
    sol = Solution()
    result = sol.balancedParanthesis(s)
    print(result)

if __name__ == "__main__":
    main()     

File: DSA/YashKrishanGupta_590014798/Day_36_Question_2.java
import java.util.*;
public class Day_36_Question_2 {
    // Solution 2
    public static boolean isBalancedParentheses(String s){
        Stack<Character> st = new Stack<>();
        for(int i=0;i<s.length();i++){
            char ch = s.charAt(i);
            if(ch=='(' || ch == '[' || ch == '{') st.push(ch);
            else if(ch == ')' || ch == ']' || ch == '}') {
                if(st.isEmpty()) return false;
                char top = st.peek();
                if((ch == ')' && top == '(') || (ch == ']' && top == '[') || (ch == '}' && top == '{')) st.pop();
                else return false;
            }
        }
        return st.isEmpty();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the String (only (,),[,],{,} ) :");
        String s = sc.nextLine();
        if(isBalancedParentheses(s)) System.out.println("Balanced");
        else System.out.println("Not Balanced");
    }
}
